subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/sourceparser/IncludeDirectivesSerializer.java

```diff
@@ -322,4 +329,47 @@ public void write(Encoder encoder, MacroFunction value) throws Exception {
             }
         }
     }
+
+    private static class CollectionSerializer<T> extends AbstractCollectionSerializer<T, Collection<T>> {
+        public CollectionSerializer(Serializer<T> entrySerializer) {
+            super(entrySerializer);
+        }
+
+        @Override
+        protected Collection<T> createCollection(int size) {
+            return new ArrayList<T>(size);
+        }
+    }
+
+    private static class MultimapSerializer<K, V> implements Serializer<Multimap<K, V>> {
```

Comment:
I wouldn't serialize the multi map, I'd reconstruct it when deserializing the macros. The problem with serialising the multi map is that the key (the macro name) is also serialised as part of the value (the macro object has a name property), so we're duplicating a lot of information on disk and also in heap. So, to serialise the macros, just iterate over the macros and serialise each one. To deserialise the macros, just read each macro object and then add to the multi map using the name from the macro object.

Reply:
That is what I thought, I will change is now.